### PR TITLE
Shutdown non-essential services before the upgrade

### PR DESCRIPTION
**Pull request by jsuchome**
Uses research from https://github.com/crowbar/crowbar-core/pull/653  and https://github.com/crowbar/crowbar-core/pull/655

### CODE REVIEW

**Comment by jsuchome** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;O26tKMj0&#x2F;98-3-s47p12-stop-unneeded-services-on-nodes-upgrade-step-10"> (3) S47P12 Stop unneeded services on nodes (upgrade step 10)</a>

File: chef/cookbooks/crowbar/recipes/prepare-upgrade-scripts.rb [hunk 1]
```diff
@@ -68,3 +70,21 @@
     new_alias: new_alias
   )
 end
+
+# This script shuts down non-essential services on the nodes
+# It leaves only database (so we can create a dump of it)
+# and services necessary for managing network traffic of running instances.
+
+# Find out now if we have HA setup and pass that info to the script
+use_ha = !search(:node, "pacemaker_founder:true").empty?
```
Comment (abel-navarro) [inline chef/cookbooks/crowbar/recipes/prepare-upgrade-scripts.rb]: this will only target one node, the founder. What about the rest?

Comment (jsuchome) [inline chef/cookbooks/crowbar/recipes/prepare-upgrade-scripts.rb]: It does not use the node for anything, it tries to find out if we are using HA setup.

But maybe this should be node specific...

File: chef/cookbooks/crowbar/recipes/prepare-services-shutdown.rb [hunk 2]
```diff
@@ -0,0 +1,34 @@
+# Cookbook Name:: crowbar
+# Recipe:: prepare-os-upgrade
```
Comment (abel-navarro) [inline chef/cookbooks/crowbar/recipes/prepare-services-shutdown.rb]: prepare-services-shutdown

Comment (jsuchome) [inline chef/cookbooks/crowbar/recipes/prepare-services-shutdown.rb]: Oh! This file should not be there, I've moved both parts to `prepare-upgrade-scripts.rb`

File: chef/cookbooks/crowbar/recipes/prepare-upgrade-scripts.rb [hunk 3]
```diff
@@ -68,3 +70,21 @@
...
 end
+
+# This script shuts down non-essential services on the nodes
+# It leaves only database (so we can create a dump of it)
+# and services necessary for managing network traffic of running instances.
+
+# Find out now if we have HA setup and pass that info to the script
+use_ha = node["run_list_map"].key? "pacemaker-cluster-member"
+
+template "/usr/sbin/crowbar-shutdown-services-before-upgrade.sh" do
```
Comment (abel-navarro) [inline chef/cookbooks/crowbar/recipes/prepare-upgrade-scripts.rb]: This looks duplicated with https://github.com/jsuchome/crowbar-core/blob/d6a4758f6a65dd40b104df90bce9eb2e6bd3c3a1/chef/cookbooks/crowbar/recipes/prepare-services-shutdown.rb#L25. What's the difference?

Comment (jsuchome) [inline chef/cookbooks/crowbar/recipes/prepare-upgrade-scripts.rb]: See last comment, prepare-services-shutdown.rb should not be there.

File: crowbar_framework/app/controllers/api/upgrade_controller.rb [hunk 4]
```diff
@@ -52,14 +52,37 @@ def prepare
...
-  api :GET, "/api/upgrade/services", "List all openstack services on all nodes that need to stop"
   api :POST, "/api/upgrade/services", "Stop related services on all nodes during upgrade"
   api_version "2.0"
   def services
-    if request.post?
-      head :not_implemented
+    status = :ok
+    msg = ""
+
+    begin
```
Comment (MaximilianMeister) [inline crowbar_framework/app/controllers/api/upgrade_controller.rb]: could you please move all this business logic into the upgrade model?

**Comment by jsuchome** [comment]
@MaximilianMeister adapted, and travis is now passing.

**Comment by abel-navarro** [comment]
+1

**Comment by MaximilianMeister** [comment]
+1 (I'll send a followup for some cleanup in #666)

**Comment by MaximilianMeister** [comment]
> (I'll send a followup for some cleanup in #666)

done

**Pull request merged**